src/marqo/core/semi_structured_vespa_index/semi_structured_add_document_handler.py

```diff
@@ -0,0 +1,74 @@
+from typing import Dict, Any
+
+from marqo.config import Config
+from marqo.core import constants
+from marqo.core.constants import MARQO_DOC_ID
+from marqo.core.document.models.add_docs_params import AddDocsParams
+from marqo.core.models.marqo_index import SemiStructuredMarqoIndex, Field, FieldType, FieldFeature, TensorField
+from marqo.core.semi_structured_vespa_index.semi_structured_vespa_index import SemiStructuredVespaIndex
+from marqo.core.semi_structured_vespa_index.semi_structured_vespa_schema import SemiStructuredVespaSchema
+from marqo.core.unstructured_vespa_index.unstructured_add_document_handler import UnstructuredAddDocumentsHandler
+from marqo.tensor_search import index_meta_cache
+from marqo.tensor_search.telemetry import RequestMetricsStore
+from marqo.vespa.models import VespaDocument
+
+
+class SemiStructuredAddDocumentsHandler(UnstructuredAddDocumentsHandler):
+    def __init__(self, marqo_index: SemiStructuredMarqoIndex, config: Config, add_docs_params: AddDocsParams):
+        super().__init__(marqo_index, config, add_docs_params)
+        self.marqo_index = marqo_index
+        self.vespa_index = SemiStructuredVespaIndex(marqo_index)
+        self.should_update_index = False
+
+    def handle_field(self, marqo_doc, field_name, field_content):
+        self._validate_field(field_name, field_content)
+        text_field_type = self._infer_field_type(field_content)
+        content = self.tensor_fields_container.collect(marqo_doc[MARQO_DOC_ID], field_name,
+                                                       field_content, text_field_type)
+        marqo_doc[field_name] = content
+
+        if isinstance(content, str):
+            # Add missing lexical fields to marqo index
+            if field_name not in self.marqo_index.field_map:
+                self.marqo_index.lexical_fields.append(
```

Comment:
This logic feels out of place here. Is there a better place? Do we need a different abstraction?

Reply:
I can create a collector class to collect fields to add to the index, but this is probably an overkill given the logic is simple and straightforward here. it's in the handle_field method since we collect the missing field along the way when we process each field of each doc.